src/helm/proxy/clients/palmyra_client.py

```diff
@@ -0,0 +1,184 @@
+import json
+import requests
+from typing import Any, Dict, List
+
+from tokenizers import Tokenizer
+from transformers import AutoTokenizer
+
+from helm.common.cache import Cache, CacheConfig
+from helm.common.hierarchical_logger import hlog
+from helm.common.request import Request, RequestResult, Sequence, Token
+from helm.common.tokenization_request import (
+    DecodeRequest,
+    DecodeRequestResult,
+    TokenizationRequest,
+    TokenizationRequestResult,
+    TokenizationToken,
+)
+from .client import Client, wrap_request_time, truncate_sequence
+from dataclasses import asdict
+
+
+class PalmyraClient(Client):
+    VALID_MODELS: List[str] = ["palmyra-base", "palmyra_large", "palmyra-r"]
+    TOKENIZER_NAME_PREFIX = "Writer/"
+
+    @staticmethod
+    def _get_model_name(model_name: str) -> str:
+        return "/".join(model_name.split("/")[1:])
+
+    def __init__(self, api_key: str, cache_config: CacheConfig):
+        self.api_key: str = api_key
+        self.cache = Cache(cache_config)
+        self._tokenizer_name_to_tokenizer: Dict[str, Tokenizer] = {}
+
+    def _get_tokenizer(self, tokenizer_name: str) -> Tokenizer:
+        modified_name = tokenizer_name.replace(self.TOKENIZER_NAME_PREFIX, "")
+        if modified_name not in self.VALID_MODELS:
+            raise ValueError(f"Invalid tokenizer: {tokenizer_name}")
+
+        # Check if the tokenizer is cached
+        if modified_name not in self._tokenizer_name_to_tokenizer:
+            self._tokenizer_name_to_tokenizer[modified_name] = AutoTokenizer.from_pretrained(
+                self.TOKENIZER_NAME_PREFIX + modified_name, use_fast=False
+            )
+            hlog(f"Initialized tokenizer: {modified_name}")
+        return self._tokenizer_name_to_tokenizer[modified_name]
+
+    def _send_request(self, model_name: str, raw_request: Dict[str, Any]) -> Dict[str, Any]:
+        response = requests.request(
+            method="POST",
+            url=f"https://enterprise-api.writer.com/llm/organization/3002/model/{model_name}/completions",
+            headers={
+                "Content-Type": "application/json",
+                "Accept": "application/json",
+                "Authorization": f"Bearer {self.api_key}",
+            },
+            data=json.dumps(raw_request),
+        )
+        result = json.loads(response.text)
+        assert "error" not in result, f"Request failed with error: {result['error']}"
```

Comment:
Raise an exception instead of asserting. Asserts are generally intended for checking internal invariants i.e. things that should always be true.

Unfortunately there are already some inappropriate asserts in the existing codebase, but we shouldn't add more.

Reply:
Ok done